sqflite/android/src/main/java/com/tekartik/sqflite/DatabaseWorkerPool.java

```diff
@@ -0,0 +1,165 @@
+ */
+public interface DatabaseWorkerPool {
+    void start();
+    void quit();
+
+    // Posts a new task.
+    //
+    // Some rules for assigning a task to a worker.
+    // - All tasks in a transaction go to the same worker. Otherwise errors will happen.
+    // - All tasks belonging to the same database run in FIFO manner. No overlapping between any
+    //   two tasks.
+    // - Tasks belonging to different databases could be run simultaneously but not necessarily
+    //   in FIFO manner.
+    void post(Database database, Runnable runnable);
+
+    static DatabaseWorkerPool create(String name, int numberOfWorkers, int priority) {
+        if (numberOfWorkers == 1) {
+            return new SingleDatabaseWorkerPoolImpl(name, priority);
+        }
+        return new DatabaseWorkerPoolImpl(name, numberOfWorkers, priority);
+    }
+}
+
+class SingleDatabaseWorkerPoolImpl implements DatabaseWorkerPool {
+
+    final String name;
+    final int priority;
+
+    private HandlerThread handlerThread;
+    private Handler handler;
+
+    SingleDatabaseWorkerPoolImpl(String name, int priority) {
+        this.name = name;
+        this.priority = priority;
+    }
+
+    @Override
+    public void start() {
+        handlerThread = new HandlerThread(name, priority);
+        handlerThread.start();
+        handler = new Handler(handlerThread.getLooper());
+    }
+
+    @Override
+    public void quit() {
+        if (handlerThread != null) {
+            handlerThread.quit();
+            handlerThread = null;
+            handler = null;
+        }
+    }
+
+    @Override
+    public void post(Database database, Runnable runnable) {
+        handler.post(runnable);
+    }
+}
+
+class DatabaseWorkerPoolImpl implements DatabaseWorkerPool {
+
+    final String name;
+    final int numberOfWorkers;
+    final int priority;
+
+    private final LinkedList<DatabaseTask> waitingList = new LinkedList<>();
+    private final LinkedList<DatabaseWorker> idleWorkers = new LinkedList<>();
+    private final LinkedList<DatabaseWorker> busyWorkers = new LinkedList<>();
+
+    DatabaseWorkerPoolImpl(String name, int numberOfWorkers, int priority) {
+        this.name = name;
+        this.numberOfWorkers = numberOfWorkers;
+        this.priority = priority;
+    }
+
+    @Override
+    public synchronized void start() {
+        for (int i = 0; i < numberOfWorkers; i++) {
+            DatabaseWorker worker = new DatabaseWorker(name + i, priority);
+            worker.start(
+                    () -> {
+                        onWorkerIdle(worker);
+                    });
+            idleWorkers.add(worker);
+        }
+    }
+
+    @Override
+    public synchronized void quit() {
+        for (DatabaseWorker worker : idleWorkers) {
+            worker.quit();
+        }
+        for (DatabaseWorker worker : busyWorkers) {
+            worker.quit();
+        }
+    }
+
+    @Override
+    public synchronized void post(Database database, Runnable runnable) {
```

Comment:
Another scenario to test is hot-reload and the weird "recoverInTransaction". Basically this happens on reload/hot reload and also when you quit the home page on Android. In this case the native part is still running. i.e. if there was a transaction in progress, the database might be lock.
The workaround done is that a ROLLBACK is sent (using transactionId = -1) and this command must be run in the same thread than where the BEGIN TRANSACTION was done previously. There is a manuel test to reproduce:
- Start the example app
- Choose 'Manual Tests'
- Tap 'openDatabase' then 'BEGIN TRANSACTION'
- restart the app from the IDE (do not terminate it)
- Choose 'Manual Tests' again
- Choose 'log level verbose' so that you see what is happening
- choose 'openDatabase'  then 'BEGIN TRANSACTION'

You'll see (when debugging) that a ROLLBACK is sent, otherwise we are stuck here and this is a common scenario if you restart your app while a transaction is in progress.
This workaround is done in the dart side as it is used in other implementation as well (cipher). Maybe you handle that correctly but it is a good (unfortunately only manual) test.

Reply:
It is an interesting scenario. I did the test and things look right (I find 'BEGIN EXCLUSIVE' but not 'BEGIN TRANSACTION' button). Here is the log. The log shows 'BEGIN EXCLUSIVE'. Thread hops (Sqflite0(74) and Sqflite1(75)) do happen and only happen between transactions.

```
D/Sqflite ( 9286): Look for /data/user/0/com.example.sqflite_test_app/databases/manual_test.db in []
D/Sqflite ( 9286): [1,flutter-worker-0(67)] starting worker pool with priority 0
D/Sqflite ( 9286): [1,flutter-worker-0(67)] opened 1 /data/user/0/com.example.sqflite_test_app/databases/manual_test.db
D/Sqflite ( 9286): [1,Sqflite0(74)] opened 1 /data/user/0/com.example.sqflite_test_app/databases/manual_test.db
D/Sqflite ( 9286): [1,Sqflite1(75)] BEGIN EXCLUSIVE
D/EGL_emulation( 9286): app_time_stats: avg=298.87ms min=1.39ms max=3155.59ms count=11

Performing hot restart...                                               
Restarted application in 687ms.
D/EGL_emulation( 9286): app_time_stats: avg=426.24ms min=1.36ms max=5474.15ms count=13
D/EGL_emulation( 9286): app_time_stats: avg=4005.66ms min=4005.66ms max=4005.66ms count=1
D/EGL_emulation( 9286): app_time_stats: avg=123.86ms min=2.04ms max=1503.73ms count=13
D/Sqflite ( 9286): Look for /data/user/0/com.example.sqflite_test_app/databases/manual_test.db in [/data/user/0/com.example.sqflite_test_app/databases/manual_test.db]
D/Sqflite ( 9286): [1,flutter-worker-0(67)] re-opened single instance (in transaction) 1 /data/user/0/com.example.sqflite_test_app/databases/manual_test.db
D/Sqflite ( 9286): [1,Sqflite1(75)] ROLLBACK
D/EGL_emulation( 9286): app_time_stats: avg=118.77ms min=1.37ms max=1435.84ms count=14
D/Sqflite ( 9286): [1,Sqflite0(74)] BEGIN EXCLUSIVE
D/EGL_emulation( 9286): app_time_stats: avg=213.68ms min=1.28ms max=2664.24ms count=13
```

Reply:
To be honest, it is hard to completely validate the logic for finding a thread without debugging which I'll do after merging but can you explain how it finds the correct thread when in a transaction (and not use another idle thread)